---
title: "miRNetR: A companion R package for the miRNet web server"
author: "Le Chang and Jianguo Xia"
date: "`r Sys.Date()`"
output:
  prettydoc::html_pretty:
    theme: 'cayman'
    highlight: github
    toc: true
    toc_depth: 3
    number_sections: true
vignette: >
  %\VignetteIndexEntry{miRNetR}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
  %\VignetteEncoding{UTF-8}
---
<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r global_options, echo = FALSE, include = FALSE}
options(width  = 999)
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE,
                      cache = FALSE, tidy = FALSE, size = "small")
```

```{r setup, echo = FALSE}
knitr::opts_chunk$set(collapse = TRUE)
```

# Overview 

## Introduction

***miRNetR*** is an R package, synchronized with the popular [miRNet](www.mirnet.ca) web server, designed for microRNA (miRNA) centric network analytics and systems-level interpretation. This R package contains the numerous R functions and libraries underlying the web server necessary to perform miRNA data processing and analysis. This package provides support to map from miRNAs to targets (forward mapping), targets to the associated miRNAs (reverse mapping), as well as to generate network files and perform functional enrichment analysis.  

Following installation and loading of *miRNetR*, users will be able to reproduce web server results from their local computers using the corresponding R command history downloaded from miRNet, thereby achieving maximum flexibility and reproducibility.

## Installation

**Step 1. Install package dependencies** 

To use miRNetR , first install all package dependencies. Ensure that you are able to download packages from bioconductor. To install package dependencies, use the pacman R package (for those with >R 3.5.1). Note that some of these packages may require additional library dependencies that need to be installed prior to their own successful installation. 

```{r, eval=FALSE}
install.packages("pacman")

library(pacman)

pacman::p_load(RSQLite, Cairo, fastmatch, igraph, RJSONIO, foreach, doParallel, preprocessCore, limma, edgeR, HTqPCR, genefilter)

```

**Step 2. Install the package**

miRNetR is freely available from GitHub. The package documentation, including the vignettes for each module and user manual is available within the downloaded R package file. If all package dependencies were installed, you will be able to install the miRNetR. Due to issues with Latex, some users may find that they are only able to install miRNetR without any documentation (i.e. vignettes). 

Install the package directly from github using the *devtools* package. Open R and enter:

```{r, eval=FALSE}

# Step 1: Install devtools
install.packages("devtools")
library(devtools)

# Step 2: Install miRNetR WITHOUT documentation
devtools::install_github("xia-lab/miRNetR", build = TRUE, build_opts = c("--no-resave-data", "--no-manual", "--no-build-vignettes"))

# Step 2: Install miRNetR WITH documentation
devtools::install_github("xia-lab/miRNetR", build = TRUE, build_opts = c("--no-resave-data", "--no-manual"), build_vignettes = TRUE)


```

# Tips for using the miRNetR package

1) The first function that you will use in every module is the `Init.Data` function, which constructs the *dataSet* object that stores user's data for further processing and analysis. 

2) The miRNetR package directly creates data files/tables/analysis/networks outputs in your current working directory. 

3) Every command must be run in sequence, please do not skip any commands as this will result in errors downstream. 

4) Each main function in miRNetR is documented. Use the *?Function* format to open its documentation. For instance, use `?miRNetR::PerformMirGeneMapping` to find out more about this function.


# Starting from a list of miRNAs or targets

## Starting from a list of miRNAs

### Mapping to single target

```{r, eval=FALSE}
library(miRNetR)

#### Step 1. Initiate the dataSet object
Init.Data("mir", "mirlist")
#> "miRNetR init done!"

#### Step 2. Set up the user input data
SetupMirListData(mirs = "hsa-mir-101-3p
hsa-mir-133b
hsa-mir-147a
hsa-mir-3140-3p
hsa-mir-361-5p
hsa-mir-510-5p", orgType = "hsa", idType = "mir_id", tissue = "Kidney")
#> "A total of  6 unique items were entered."

#### Step 3. Perform miRNAs to target genes mapping, 
#### results are downloaded in your working directory ("mirnet_mir_target.csv" and "mir_tissue.csv")
PerformMirGeneMapping()
#> "Downloading mir2gene.sqlite from https://www.xialab.ca/resources/mirnet/mir2gene.sqlite"
#> "A total of unqiue 1059 pairs of miRNA-gene targets were identified!"

head(dataSet$mir.res, n = 3L)
#>                             ID    Accession   Gene Entrez           Experiment        Literature Tissue
#> mirnet-hsa-21491 hsa-mir-101-3p MIMAT0000099 ACVR2B     93           Sequencing          20371350 Kidney
#> mirnet-hsa-21492 hsa-mir-101-3p MIMAT0000099  AP1G1    164 PAR-CLIP//Sequencing 20371350|23592263 Kidney
#> mirnet-hsa-21493 hsa-mir-101-3p MIMAT0000099   AMD1    262            HITS-CLIP          23313552 Kidney

```


```{r, eval=FALSE}
#### Step 4. Generate miRNA-gene network files
CreateMirNets(net.type = "mir2gene")
#> 6    6  999 1059    1 1005
## mir.query = 6, mir.count = 6, tgt.count = 999, ecount(mir.graph) = 1059, length(mir.nets) = 1, substats = 1005;

#### Step 5. Prepare network files, 
#### results are downloaded in your working directory ("node_table_mirnet_0.csv", "mirnet_0.json" and "mirnet.graphml")
PrepareMirNet(mir.nm = "mirnet1", file.nm = "mirnet_0.json")

```

```{r, eval=FALSE}
#### Step 6. Perform miRNA family enrichment analysis, 
#### results are downloaded in your working directory ("network_enrichment_mirfamily_1.json" and "mirnet_enrichment.csv")
PerformMirTargetEnrichAnalysis(
  adjust.type = "NA",
  fun.type = "mirfamily",
  file.nm = "network_enrichment_mirfamily_1",
  IDs = "hsa-mir-101-3p; hsa-mir-147a; hsa-mir-361-5p; hsa-mir-133b; hsa-mir-510-5p; hsa-mir-3140-3p",
  algo = "hyp"
)

resTable <- read.csv("mirnet_enrichment.csv", header=T, as.is=T)
head(resTable, n = 3L)
#>    Pathway Total Expected Hits    Pval stringsAsFactor
#>  1 mir-3140     3  0.00753    1 0.00751           FALSE
#>  2  mir-361     3  0.00753    1 0.00751           FALSE
#>  3  mir-147     4  0.01000    1 0.01000           FALSE
```

### Mapping to multiple targets

```{r, eval=FALSE}
library(miRNetR)

#### Step 1. Initiate the dataSet object
Init.Data("mir", "mirlist")

#### Step 2. Set up the user input data
SetupMirListData(mirs = "hsa-mir-16-5p
hsa-mir-424-5p
hsa-mir-29c-3p
hsa-mir-29b-3p
hsa-mir-29a-3p
hsa-mir-148b-3p
hsa-mir-152-3p
hsa-mir-148a-3p
hsa-mir-335-5p", orgType = "hsa", idType = "mir_id", tissue = "na")

#### Step 3. Set up targets
nms.vec = c("gene", "lncrna") 
SetCurrentDataMulti()

#### Step 4. Perform miRNAs to multiple targets mapping, results are downloaded in your working directory
QueryMultiListMir()
head(dataSet$mir2gene, n = 2L)
#>              ID    Accession Target TargetID Experiment Literature      Tissue
#> 1 hsa-mir-16-5p MIMAT0000069  ABCB7       22      CLASH   23622248 Unspecified
#> 2 hsa-mir-16-5p MIMAT0000069  ABCF1       23 Proteomics   18668040 Unspecified
head(dataSet$mir2lnc, n = 2L)
#>                ID    Accession     Target TargetID Experiment Literature      Tissue
#> 1 hsa-mir-148a-3p MIMAT0000243      SNHG3     8420   CLIP-Seq   24297251 Unspecified
#> 2 hsa-mir-148a-3p MIMAT0000243 CYP1B1-AS1   285154   CLIP-Seq   24297251 Unspecified
```

```{r, eval=FALSE}
#### Step 5. Generate network files
CreateMirNets(net.type = "multilist")
#> 18    9 4848 6752    1 4857
## mir.query = 18, mir.count = 9, tgt.count = 4848, ecount(mir.graph) = 6752, length(mir.nets) = 1, substats = 4857;

#### Step 6. Prepare network files, 
#### results are downloaded in your working directory ("node_table_mirnet_0.csv", "mirnet_0.json" and "mirnet.graphml")
PrepareMirNet(mir.nm = "mirnet1", file.nm = "mirnet_0.json")
```

```{r, eval=FALSE}
#### Step 7. Perform miRNA function enrichment analysis, 
#### results are downloaded in your working directory ("network_enrichment_func_1.json" and "mirnet_enrichment.csv")
PerformMirTargetEnrichAnalysis(
  adjust.type = "NA",
  fun.type = "func",
  file.nm = "network_enrichment_func_1",
  IDs = "hsa-mir-16-5p; hsa-mir-424-5p; hsa-mir-29c-3p; hsa-mir-29b-3p; hsa-mir-29a-3p; hsa-mir-148b-3p; hsa-mir-152-3p; hsa-mir-148a-3p; hsa-mir-335-5p",
  algo = "hyp"
)

resTable <- read.csv("mirnet_enrichment.csv", header=T, as.is=T)
head(resTable, n = 3L)
#>            Pathway Total Expected Hits    Pval stringsAsFactor
#> 1    Cell Adhesion    24    0.228    3 0.00109           FALSE
#> 2 Neuron Apoptosis    45    0.428    3 0.00692           FALSE
#> 3   Nephrotoxicity    51    0.485    3 0.00986           FALSE
```

## Starting from a list of targets

### Mapping from single target

```{r message = FALSE,warning = FALSE}
library(miRNetR)

#### Step 1. Initiate the dataSet object
Init.Data("mir", "gene")
#> "miRNetR init done!"

#### Step 2. Set up the user input data
SetupIndListData("ACAT1
DDX3X
MEGF9
FMR1
HAPLN1
IL12A
OLR1
RNF6
SOD3
ELOVL4
MRAP2", "hsa", "gene", "symbol", "na", "trrust");
#> "A total of  11 unique items were entered."

#### Step 3. Perform target gene to miRNA mapping, 
#### results are downloaded in your working directory ("mirnet_mir_target.csv")
QueryMultiList()
#> "Downloading mir2gene.sqlite from https://www.xialab.ca/resources/mirnet/mir2gene.sqlite"
#> "A total of unqiue 1059 pairs of miRNA-gene targets were identified!"

head(dataSet$mir.res, n = 3L)
#>                             ID    Accession   Gene Entrez           Experiment        Literature Tissue
#> mirnet-hsa-21491 hsa-mir-101-3p MIMAT0000099 ACVR2B     93           Sequencing          20371350 Kidney
#> mirnet-hsa-21492 hsa-mir-101-3p MIMAT0000099  AP1G1    164 PAR-CLIP//Sequencing 20371350|23592263 Kidney
#> mirnet-hsa-21493 hsa-mir-101-3p MIMAT0000099   AMD1    262            HITS-CLIP          23313552 Kidney

```

### Mapping from multiple targets

```{r message = FALSE,warning = FALSE}
library(miRNetR)

#### Step 1. Mapping

#### Step 2. Network generation

#### Step 3. Enrichment analysis (hypergeometric tests)

#### Step 3. Enrichment analysis (empirical sampling)

#### Step 3. Enrichment analysis (miRNA set enrichment)
```

# Starting from an expression table 

## RT-qPCR

```{r, eval=FALSE}
library(miRNetR)

#### Step 1. Initiate the dataSet object
Init.Data("mir", "qpcr")

#### Step 2. Read user input 
ReadTabExpressData("qpcr_data.txt")

#### Step 3. Annotation
PerformDataAnnot("mmu", "mir_id",  "na", "mean")
```
miRNetR supports five types of normalization method, including quantile normalization, rank-invariant normalization, acale-invariant normalization, geometric average and delta Ct normalization.

`Quantile normalization` will make the distribution of control values more or less identical across samples.

`Rank-invariant normalization` computes all rank-invariant sets of features between pairwise comparisons of each sample against a reference, such as a pseudo-mean. The rank-invariant features are used as a reference for generating a smoothing curve, which is then applied to the entire sample.

`Geometric average` calculates the average control value for each sample, and scales all control values according to the ratio of these mean Ct values across samples. 

`deltaCt normalization` calculates the standard deltaCt values, i.e. subtracts the mean of the chosen controls from all other values in the feature set. *You must specify endogenous controls!*


```{r, eval=FALSE}
#### Step 4. Normalization

######## Option 1. Quantile normalization
PerformQpcrDataNormalization("quantile")

######## Option 2. Rank-invariant normalization
PerformQpcrDataNormalization("norm.rank")

######## Option 3. Scale-invariant normalization
PerformQpcrDataNormalization("scale.rank")

######## Option 4. Geometric average
PerformQpcrDataNormalization("geometric.mean")

######## Option 5. Delta Ct normalization
PerformQpcrDataNormalization("deltaCt")

######## Option 6. No normalization
PerformQpcrDataNormalization("none")
```
miRNetR supports three types of statistical method, including limma, standard t-test and Mann-Whitney test.
```{r, eval=FALSE}
#### Step 5. Specify comparison of interest and statistical method

######## Option 1. Limma
PerformHTqPCR("Ctrl vs. GrpA", "limma")

######## Option 2. Standard t-test
PerformHTqPCR("Ctrl vs. GrpA", "ttest")

######## Option 3. Mann-Whitney test
PerformHTqPCR("Ctrl vs. GrpA", "mannwhitney")
```
Set adjusted p-value (p.lvl), fold change (fc.lvl) and direction
```{r, eval=FALSE}
#### Step 6. Feature selection

######## Option 1. Both direction
GetSigGenes(p.lvl = 0.2, fc.lvl = 0.0, direction = "both")

######## Option 2. Up-regulated only
GetSigGenes(p.lvl = 0.2, fc.lvl = 0.0, direction = "up")

######## Option 3. Down-regulated only
GetSigGenes(p.lvl = 0.2, fc.lvl = 0.0, direction = "down")
```

```{r, eval=FALSE}
#### Step 7. Set up the user input data in the dataSet object
SetupMirExpressData();

#### Step 8. Perform miRNAs to target genes mapping, results are downloaded in your working directory
PerformMirGeneMapping();
```

```{r, eval=FALSE}
#### Step 9. Generate miRNA-gene network files
CreateMirNets("mir2gene");

#### Step 10. Prepare network files, results are downloaded in your working directory
PrepareMirNet("mirnet1", "mirnet_0.json")
## "node_table_mirnet_0.csv", "mirnet_0.json" and "mirnet.graphml" are downloaded in your working directory 

```

```{r, eval=FALSE}
#### Step 11. Perform miRNA family enrichment analysis, results are downloaded in your working directory ("network_enrichment_mirfamily_1.json" and "mirnet_enrichment.csv")
PerformMirTargetEnrichAnalysis(
  adjust.type = "NA",
  fun.type = "mirfamily",
  file.nm = "network_enrichment_mirfamily_1",
  IDs = "mmu-mir-329-3p; mmu-mir-539-3p; mmu-mir-3087-3p; mmu-mir-466i-5p; mmu-mir-129-2-3p; mmu-mir-466c-3p; mmu-mir-669m-3p; mmu-mir-466e-3p; mmu-mir-325-3p; mmu-mir-3099-3p; mmu-mir-1952; mmu-mir-505-3p; mmu-mir-451a",
  algo = "hyp"
)
```

## RNAseq

```{r, eval=FALSE}
library(miRNetR)

#### Step 1. Initiate the dataSet object
Init.Data("mir", "rnaseq")

#### Step 2. Read user input 
ReadTabExpressData("bmdm_data.txt")

#### Step 3. Annotation
PerformDataAnnot("mmu", "emblgene",  "na", "sum")
```
miRNetR supports four types of normalization method, including trimmed mean of M-values (TMM), log2 transformation only, quantile normalization only, log2 followed by quantile normalization. In addition, you can choose either tagwise or common dispersion. The expression values should be compared at log2 scales.
```{r, eval=FALSE}
#### Step 4. Normalization

######## Option 1. Trimmed mean of M-values (TMM) and tagwise dispersion
PerformCountDataNormalization("tmm", "tagwise")

######## Option 2. Log2 transformation and common dispersion
PerformCountDataNormalization("log", "common")

######## Option 3. Quantile normalization and common dispersion
PerformCountDataNormalization("quant", "common")

######## Option 4. Log2 followed by quantile normalization and common dispersion
PerformCountDataNormalization("combine", "common")

######## Option 5. No normalization and no dispersion
PerformCountDataNormalization("none", "none")
```
EdgeR uses negative binomial model is to estimate the dispersion parameter for each tag, a measure of the degree of inter-library variation for that tag. 
Estimating the common dispersion gives an idea of overall variability across the genome for this dataset.
Estimating the tagwise dispersions estimates a distinct, individual dispersion for each gene.

```{r, eval=FALSE}
#### Step 5. Specify statistical method and comparison of interest

######## edgeR (RNAseq)
PerformEdgeR("Control vs. Infected")

```
Set adjusted p-value (p.lvl), fold change (fc.lvl) and direction
```{r, eval=FALSE}
#### Step 6. Feature selection

######## Option 1. Both direction
GetSigGenes(p.lvl = 0.01, fc.lvl = 2.5, direction = "both")

######## Option 2. Up-regulated only
GetSigGenes(p.lvl = 0.01, fc.lvl = 2.5, direction = "up")

######## Option 3. Down-regulated only
GetSigGenes(p.lvl = 0.01, fc.lvl = 2.5, direction = "down")
```

```{r, eval=FALSE}
#### Step 7. Set up the user input data in the dataSet object
SetupMirExpressData();

#### Step 8. Perform target genes to miRNAs mapping, results are downloaded in your working directory
PerformMirGeneMapping();
```

```{r, eval=FALSE}
#### Step 9. Generate miRNA-gene network files
CreateMirNets("gene2mir");

#### Step 10. Prepare network files, results are downloaded in your working directory
PrepareMirNet("mirnet1", "mirnet_0.json")
## "node_table_mirnet_0.csv", "mirnet_0.json" and "mirnet.graphml" are downloaded in your working directory 

```

```{r, eval=FALSE}
#### Step 11. Perform miRNA family enrichment analysis, results are downloaded in your working directory ("network_enrichment_mirfamily_1.json" and "mirnet_enrichment.csv")
PerformMirTargetEnrichAnalysis(
  adjust.type = "NA",
  fun.type = "mirfamily",
  file.nm = "network_enrichment_mirfamily_1",
  IDs = "mmu-mir-329-3p; mmu-mir-539-3p; mmu-mir-3087-3p; mmu-mir-466i-5p; mmu-mir-129-2-3p; mmu-mir-466c-3p; mmu-mir-669m-3p; mmu-mir-466e-3p; mmu-mir-325-3p; mmu-mir-3099-3p; mmu-mir-1952; mmu-mir-505-3p; mmu-mir-451a",
  algo = "hyp"
)
```

## Microarray

```{r, eval=FALSE}
library(miRNetR)

#### Step 1. Initiate the dataSet object
Init.Data("mir", "array")

#### Step 2. Read user input 
ReadTabExpressData("estrogen_data.txt");

#### Step 3. Annotation
PerformDataAnnot("hsa", "hgu95av2",  "na", "mean");
```
miRNetR supports four types of normalization method, including trimmed mean of M-values (TMM), log2 transformation only, quantile normalization only, log2 followed by quantile normalization. In addition, you can choose either tagwise or common dispersion.
The expression values should be compared at log2 scales. Quantile normalization is optional and is recommended for microarray data. 

```{r, eval=FALSE}
#### Step 4. Normalization

######## Option 1. Trimmed mean of M-values (TMM) 
PerformArrayDataNormalization("tmm")

######## Option 2. Log2 transformation 
PerformArrayDataNormalization("log")

######## Option 3. Quantile normalization 
PerformArrayDataNormalization("quant")

######## Option 4. Log2 followed by quantile normalization 
PerformArrayDataNormalization("combine")

######## Option 5. No normalization 
PerformArrayDataNormalization("none")

```

```{r, eval=FALSE}
#### Step 5. Specify statistical method and comparison of interest

######## Limma (microarray)
PerformLimma("absent vs. present");

```
Set adjusted p-value (p.lvl), fold change (fc.lvl) and direction
```{r, eval=FALSE}
#### Step 6. Feature selection

######## Option 1. Both direction
GetSigGenes(p.lvl = 0.01, fc.lvl = 1.0, direction = "both")

######## Option 2. Up-regulated only
GetSigGenes(p.lvl = 0.01, fc.lvl = 1.0, direction = "up")

######## Option 3. Down-regulated only
GetSigGenes(p.lvl = 0.01, fc.lvl = 1.0, direction = "down")
```

```{r, eval=FALSE}
#### Step 7. Set up the user input data in the dataSet object
SetupMirExpressData();

#### Step 8. #### Step 8. Perform target genes to miRNAs mapping, results are downloaded in your working directory
PerformMirGeneMapping();
```

```{r, eval=FALSE}
#### Step 9. Generate miRNA-gene network files
CreateMirNets("gene2mir");

#### Step 10. Prepare network files, results are downloaded in your working directory
PrepareMirNet("mirnet1", "mirnet_0.json")
## "node_table_mirnet_0.csv", "mirnet_0.json" and "mirnet.graphml" are downloaded in your working directory 

```

```{r, eval=FALSE}
#### Step 11. Perform miRNA family enrichment analysis, results are downloaded in your working directory ("network_enrichment_mirfamily_1.json" and "mirnet_enrichment.csv")
PerformMirTargetEnrichAnalysis(
  adjust.type = "NA",
  fun.type = "mirfamily",
  file.nm = "network_enrichment_mirfamily_1",
  IDs = "mmu-mir-329-3p; mmu-mir-539-3p; mmu-mir-3087-3p; mmu-mir-466i-5p; mmu-mir-129-2-3p; mmu-mir-466c-3p; mmu-mir-669m-3p; mmu-mir-466e-3p; mmu-mir-325-3p; mmu-mir-3099-3p; mmu-mir-1952; mmu-mir-505-3p; mmu-mir-451a",
  algo = "hyp"
)
```

# Starting from a list of xeno-miRNAs or target genes

miRNet will automatically search 453 xeno-miRNAs detected in six host organisms based on 15 studies. In addition, there are 1,249 predicted xeno-miRNAs with high potential to be transferred to human circulation according to a study by [Shu et al.](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4631372/)

You can enter xeno-miRNAs to see if they target any host genes; Or enter host genes to see if they are targeted by any xeno-miRNAs.

## Starting from a list of xeno-miRNAs

```{r, eval=FALSE}
#### Step 1. Initiate the dataSet object
Init.Data("xeno.mir", "xenomir-search")

#### Step 2. Set up the user input data
SetupMirListData(mirs = "sja-miR-125b
sja-miR-2162-3p
sja-miR-2b-5p
sja-miR-61
sja-miR-10-5p", orgType = "hsa", idType = "exo_mirna", tissue = "")

#### Step 3. Perform xeno-miRNAs to target genes mapping, results are downloaded in your working directory
## You can choose to include ("true") or exclude ("false") predicted xeno-miRNAs in PerformXenoMirGeneMapping()
PerformXenoMirGeneMapping(isIncludePre = "false")
```



```{r, eval=FALSE}
#### Step 4. Generate xeno-miRNA gene network files
CreateMirNets(net.type = "mir2gene")

#### Step 5. Prepare network files, results are downloaded in your working directory
PrepareMirNet(mir.nm = "mirnet1", file.nm = "mirnet_0.json")
## "node_table_mirnet_0.csv", "mirnet_0.json" and "mirnet.graphml" are downloaded in your working directory 
```

```{r, eval=FALSE}
#### Step 6. Perform enrichment analysis, results are downloaded in your working directory ("network_enrichment_mirfamily_1.json" and "mirnet_enrichment.csv")
## TO-DO
PerformMirTargetEnrichAnalysis(
  adjust.type = "NA",
  fun.type = "mirfamily",
  file.nm = "network_enrichment_mirfamily_1",
  IDs = "hsa-mir-101-3p; hsa-mir-147a; hsa-mir-361-5p; hsa-mir-133b; hsa-mir-510-5p; hsa-mir-3140-3p",
  algo = "hyp"
)
```

## Starting from a list of host genes

```{r, eval=FALSE}
#### Step 1. Initiate the dataSet object
Init.Data("xeno.mir", "xenomir-search")

#### Step 2. Set up the user input data
SetupMirListData(mirs = "B4galt2
Acox3
Emp2
Gtpbp2
Duox1", orgType = "mmu", idType = "symbol", tissue = "")

#### Step 3. Perform xeno-miRNAs to target genes mapping, results are downloaded in your working directory
## You can choose to include ("true") or exclude ("false") predicted xeno-miRNAs in PerformXenoMirGeneMapping()
PerformXenoMirGeneMapping(isIncludePre = "false")
```

```{r, eval=FALSE}
#### Step 4. Generate xeno-miRNA gene network files
CreateMirNets(net.type = "mir2gene")

#### Step 5. Prepare network files, results are downloaded in your working directory
PrepareMirNet(mir.nm = "mirnet1", file.nm = "mirnet_0.json")
## "node_table_mirnet_0.csv", "mirnet_0.json" and "mirnet.graphml" are downloaded in your working directory 
```

```{r, eval=FALSE}
#### Step 6. Perform enrichment analysis, results are downloaded in your working directory ("network_enrichment_mirfamily_1.json" and "mirnet_enrichment.csv")
PerformMirTargetEnrichAnalysis(
  adjust.type = "NA",
  fun.type = "kegg",
  file.nm = "network_enrichment_kegg_1",
  IDs = "bta-miR-6120-3p; cel-miR-229-5p; cel-miR-237-5p; cel-miR-35-3p; cel-miR-40-3p; cel-miR-47-5p; cel-miR-53-5p; cel-miR-55-3p; cel-miR-64-5p; cel-miR-66-5p; cel-miR-81-5p; cpa-miR166d; cpa-miR167d; cqu-miR-278; cqu-miR-281-5p; dme-miR-996-3p; dya-miR-286; dya-miR-2b; dya-miR-2c; dya-miR-316; dya-miR-3; lgi-miR-96a; lus-miR166j; mdo-miR-24-5p; mse-miR-263a; mse-miR-275; nvi-miR-279; nvi-miR-2a; pab-miR950-5p; sof-miR168b; ssc-miR-432-5p; ssp-miR168a; Emp2; Gtpbp2; B4galt2; Acox3",
  algo = "hyp"
)
```

# mRNA-miRNA co-expression network analysis (To-Do)

# Network motif analysis (To-Do)


